aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move

```diff
@@ -25,23 +26,41 @@ module aptos_framework::gas_schedule {
         entries: vector<GasEntry>
     }
 
+    struct GasScheduleV2 has key, copy, drop {
+        feature_version: u64,
+        entries: vector<GasEntry>,
+    }
+
     /// Only called during genesis.
     public(friend) fun initialize(aptos_framework: &signer, gas_schedule_blob: vector<u8>) {
         system_addresses::assert_aptos_framework(aptos_framework);
         assert!(!vector::is_empty(&gas_schedule_blob), error::invalid_argument(EINVALID_GAS_SCHEDULE));
 
         // TODO(Gas): check if gas schedule is consistent
-        move_to<GasSchedule>(aptos_framework, from_bytes(gas_schedule_blob));
+        let gas_schedule: GasScheduleV2 = from_bytes(gas_schedule_blob);
+        move_to<GasScheduleV2>(aptos_framework, gas_schedule);
     }
 
-    /// This can be called by on-chain governance to update gas schedule.
-    public fun set_gas_schedule(aptos_framework: &signer, gas_schedule_blob: vector<u8>) acquires GasSchedule {
+    /// This can be called by on-chain governance to update the gas schedule.
+    public fun set_gas_schedule(aptos_framework: &signer, gas_schedule_blob: vector<u8>) acquires GasSchedule, GasScheduleV2 {
         system_addresses::assert_aptos_framework(aptos_framework);
         assert!(!vector::is_empty(&gas_schedule_blob), error::invalid_argument(EINVALID_GAS_SCHEDULE));
 
-        // TODO(Gas): check if gas schedule is consistent
-        let gas_schedule = borrow_global_mut<GasSchedule>(@aptos_framework);
-        *gas_schedule = from_bytes(gas_schedule_blob);
+        if (exists<GasScheduleV2>(@aptos_framework)) {
+            let gas_schedule = borrow_global_mut<GasScheduleV2>(@aptos_framework);
+            let new_gas_schedule: GasScheduleV2 = from_bytes(gas_schedule_blob);
+            assert!(new_gas_schedule.feature_version >= gas_schedule.feature_version, EINVALID_GAS_FEATURE_VERSION);
+            // TODO(Gas): check if gas schedule is consistent
+            *gas_schedule = new_gas_schedule;
+        }
+        else {
+            if (exists<GasSchedule>(@aptos_framework)) {
+                _ = move_from<GasSchedule>(@aptos_framework);
+            };
+            let new_gas_schedule: GasScheduleV2 = from_bytes(gas_schedule_blob);
+            // TODO(Gas): check if gas schedule is consistent
+            move_to<GasScheduleV2>(aptos_framework, new_gas_schedule);
+        };
```

Comment:
@movekevin @wrwg this currently implements a lazy-upgrade scheme: when we set a new gas schedule, it's gonna check if there exists a V1 version of the gas schedule and delete it if so.

However once caveat is that, once we deploy the upgraded modules, we cannot update the gas schedule before all validators have deployed the new Rust binaries, or otherwise the ones running the old software will be in trouble

Reply:
We need to carefully engineer the sequence in how we upgrade testnet... @davidiw @movekevin